### PR TITLE
Allow upload of dbfs at national level

### LINKED ISSUE #452: Atualizar o APP de upload de DBFs para aceitar DBFs do Brasil
Atualmente o formulário de upload de DBFs, requer a especificação do estado a que se refere o DBF. Porém, agora que o Infodengue recebe DBF de todo o país, podemos remover este campo ou permitir a entrada `BR` caso seja útil.

Outra falha é que na rotina de validação  do modelo de dados o processo falha se o campo `NM_BAIRRO` não existe. Este erro deve ser convertido em um simples warning, pois o DBF nacional não o contém.
> BR-CHIK-2021-08-23-2021.dbf
nm_bairro

### PR DESCRIPTION
**Pull request by esloch**
- Resolves #452
- Removed dump files from staging_db container to use as volume in compose file pointing to filesystem (reduced image)

> - Inserted In the list of states option to select the dbf at the national level Brazil-BR.

> - Changed the state list name from *state_abbreviation* to **ABREVIATION** in *models.py*

> - Removed **ID_BAIRRO** and **NM_BAIRRO** fields from expected_fields list to allow dbfs input without these fields.

> - Run the migration in the **infodengue** database to update the column name to **abbreviation** in the table **"Municipio"."Notification"**

> - Modified the timeout in *NGINX* so as not to give an error when loading the DBF file.

> - Modified gunicorn timeout so as not to reset the worker's time when entering the database

> - Changed the *dbfread* lib to *geopandas* to allow opening big data files

> - Created a *DOCKER_HOST_PQDIR* volume in the compose file pointing to /Storage to allow writing and reading the files (.parquet) from the container

> - Created [utils](https://github.com/AlertaDengue/AlertaDengue/blob/962db2c3dc668142595cb5969ee16f82f619dfc1/AlertaDengue/dbf/utils.py) module with **chunk_gen** and **chunk_dbf_toparquet** functions to save and load files in chunks from filesystem in parquet format

> - Introduced the lib *simpledbf* in the code to use the parameters as variables.

> - Passed the *FIELD_MAP* dictionary to utils.py module for use in *sinan.py*

> - Fixed typing error when values ​​are *NaN* in the **ID_BAIRRO** field (row[7]) on the sinan.py module

#### utility:
``` 
time python AlertaDengue/manage.py load_sinan /tmp/DBFS/Dengue_SE30.dbf 2021
```

### CODE REVIEW

**Comment by esloch** [comment]
Changed *U.F* label to *Abrangência*:

chik_se33_form_20210830
--- 
Inserted *BR-Brasil* for selection in *ABBREVIATION_CODE_CHOICES*
```
ABBREVIATION_CODE_CHOICES = (
    ('BR', 'Brasil'),
    (None, '-----'),
    ('AC', 'Acre'),
    ('AL', 'Alagoas'),
    ('AP', 'Amapá'),
    ('AM', 'Amazonas'),
...
```
--- 
Table dbf_dbf result:

table_dbf_dbf_upload_20210830

@fccoelho do you think? Change the column name from *state_abbreviation* to *abbreviation_code_choices*?

```
infodengue> \d dbf_dbf
+--------------------+--------------------------+-------------------------------------------------------+
| Column             | Type                     | Modifiers                                             |
|--------------------+--------------------------+-------------------------------------------------------|
| id                 | integer                  |  not null default nextval('dbf_dbf_id_seq'::regclass) |
| file               | character varying(100)   |  not null                                             |
| uploaded_at        | timestamp with time zone |  not null                                             |
| export_date        | date                     |  not null                                             |
| notification_year  | integer                  |  not null                                             |
| uploaded_by_id     | integer                  |  not null                                             |
| state_abbreviation | character varying(2)     |                                                       |
| municipio          | character varying(255)   |  not null                                             |
+--------------------+--------------------------+-------------------------------------------------------+
```

**Comment by fccoelho** [comment]
@esloch, let's keep it simple and just call it `abbreviation`

File: AlertaDengue/dbf/forms.py [hunk 1]
```diff
@@ -20,7 +19,7 @@ class DBFForm(forms.Form):
     export_date = forms.DateField(label=_("Data da exportação"))
     notification_year = forms.IntegerField(label=_("Ano de notificação"))
     state_abbreviation = forms.ChoiceField(
-        choices=DBF.STATE_ABBREVIATION_CHOICES, label=_("U.F.")
+        choices=DBF.ABBREVIATION_CODE_CHOICES, label=_("Abrangência")
```
Comment (fccoelho) [inline AlertaDengue/dbf/forms.py:22]: call it just `abbreviation`

**Review by fccoelho — CHANGES_REQUESTED** [review]
just change the column name.

**Comment by esloch** [comment]
> just change the column name.

```
infodengue> \d dbf_dbf
+-------------------+--------------------------+-------------------------------------------------------+
| Column            | Type                     | Modifiers                                             |
|-------------------+--------------------------+-------------------------------------------------------|
| id                | integer                  |  not null default nextval('dbf_dbf_id_seq'::regclass) |
| file              | character varying(100)   |  not null                                             |
| uploaded_at       | timestamp with time zone |  not null                                             |
| export_date       | date                     |  not null                                             |
| notification_year | integer                  |  not null                                             |
| uploaded_by_id    | integer                  |  not null                                             |
| abbreviation      | character varying(2)     |                                                       |
| municipio         | character varying(255)   |  not null                                             |
+-------------------+--------------------------+-------------------------------------------------------+

```


> just change the column name.

Done!
```
infodengue> \d dbf_dbf
+-------------------+--------------------------+-------------------------------------------------------+
| Column            | Type                     | Modifiers                                             |
|-------------------+--------------------------+-------------------------------------------------------|
| id                | integer                  |  not null default nextval('dbf_dbf_id_seq'::regclass) |
| file              | character varying(100)   |  not null                                             |
| uploaded_at       | timestamp with time zone |  not null                                             |
| export_date       | date                     |  not null                                             |
| notification_year | integer                  |  not null                                             |
| uploaded_by_id    | integer                  |  not null                                             |
| abbreviation      | character varying(2)     |                                                       |
| municipio         | character varying(255)   |  not null                                             |
+-------------------+--------------------------+-------------------------------------------------------+

```

**Comment by esloch** [comment]
> @esloch, let's keep it simple and just call it `abbreviation`  

Done!
```
class DBF(models.Model):
    ABBREVIATION = (
        ('BR', 'Brasil'),
        (None, '-----'),
        ('AC', 'Acre'),
        ('AL', 'Alagoas'),
        ('AP', 'Amapá'),
        ('AM', 'Amazonas'),
 ```

**Comment by esloch** [comment]
> just change the column name.

Used the *migrations.RenameField* according to [this tip](https://gist.github.com/dhbradshaw/e2bdeb502b0d0d2acced)

**Comment by esloch** [comment]
@fccoelho I did some tests on _staging, NGINX is giving connection timeout error when uploading dbf:
```
2021/09/01 12:19:56 [error] 54402#54402: *1114270 upstream prematurely closed connection while reading response  
header from upstream, client: 179.190.120.154, server: infostaging.dengue.mat.br, request:   
"POST /dbf/upload/ HTTP/1.1", upstream: "http://127.0.0.1:8080/dbf/upload/",   
host: "infostaging.dengue.mat.br", referrer: "https://infostaging.dengue.mat.br/dbf/upload/"

179.190.120.154 - - [01/Sep/2021:12:19:56 -0300] "GET /favicon.ico HTTP/1.1"   
404 3303 "https://infostaging.dengue.mat.br/dbf/upload/" "Mozilla/5.0 (X11; Linux x86_64)   
AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.159 Safari/537.36"
``` 
I will do some tests in nginx on the server: 
https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_read_timeout

**Comment by fccoelho** [comment]
maybe this can help:
https://serverfault.com/questions/414987/nginx-proxy-timeout-while-uploading-big-files

**Comment by esloch** [comment]
> maybe this can help:
> https://serverfault.com/questions/414987/nginx-proxy-timeout-while-uploading-big-files

Thanks!

**Comment by esloch** [comment]
@fccoelho I modified NGINX like this:
```
    location / {
        proxy_pass http://127.0.0.1:8080;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_connect_timeout      1200;
        proxy_send_timeout         1200;
        proxy_read_timeout         1200;
        send_timeout               1200;
        client_max_body_size       10m;
        client_body_buffer_size    128k;        
    }
  ```
and gunicorn timeout to::
  ```
  ad_main.wsgi:application --timeout 1200
  ``` 
I got this result:
  ```
  ...Instanciando SINAN (/MEDIA_ROOT/BR-DEN-2021-08-24-2021_BqKjgjs.dbf, 2021)
.....Task dbf.tasks.import_dbf_to_database[21e0cd70-be3c-44ee-94c8-7a969d086b73] raised unexpected: MemoryError((784130, 123), dtype('O'))
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/celery/app/trace.py", line 450, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/srv/deploy/AlertaDengue/dbf/tasks.py", line 75, in import_dbf_to_database
    sinan = Sinan(dbf.file.path, dbf.notification_year)
  File "/srv/deploy/AlertaDengue/dbf/sinan.py", line 86, in __init__
    self.tabela = pd.DataFrame(self.dbf.records)
  File "/opt/conda/lib/python3.8/site-packages/pandas/core/frame.py", line 509, in __init__
    arrays, columns = to_arrays(data, columns, dtype=dtype)
  File "/opt/conda/lib/python3.8/site-packages/pandas/core/internals/construction.py", line 526, in to_arrays
    return _list_of_dict_to_arrays(
  File "/opt/conda/lib/python3.8/site-packages/pandas/core/internals/construction.py", line 645, in _list_of_dict_to_arrays
    content = list(lib.dicts_to_array(data, list(columns)).T)
  File "pandas/_libs/lib.pyx", line 379, in pandas._libs.lib.dicts_to_array
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 736. MiB for an array with shape (784130, 123) and data type object
  ```
  Here, we came across what we had said about in the meeting...

**Comment by fccoelho** [comment]
Memory issue. Like we discussed, we have to change the code to read from the DBF in chunks.

**Comment by esloch** [comment]
@fccoelho we will have to open a issue to change from dbfread to geopandas in the other functions :)
profiling: 
Screenshot_20210914_205823

**Comment by fccoelho** [comment]
Why geopandas and not just pandas?

**Comment by esloch** [comment]
> Why geopandas and not just pandas?  

*import geopandas as gpd...*  

What do you think if we pass that on in the views?
https://github.com/AlertaDengue/AlertaDengue/blob/main/AlertaDengue/dbf/views.py#L40-L41

I'm going to create utils module for dbf:
``` 
def chunk_gen(chunksize, totalsize):
    chunks = totalsize//chunksize

    for i in range(chunks):
        yield i * chunksize, (i + 1) * chunksize
    
    rest = totalsize % chunksize
    
    if rest:
        yield (
            chunks * chunksize, 
            (chunks * chunksize) + rest
        )

def chunk_dbf_filter(dbfname):
    dbf = Dbf5(dbfname)
    for chunk, (lowerbound, upperbound) in enumerate(
        chunk_gen(1000, dbf.numrec)
    ):
        df = gpd.read_file(
            dbfname, rows=slice(lowerbound, upperbound)
        )
        df[field_map.values()].to_parquet(f"./DBFS/data/BR_ZIKA_SE30_2021/data-{chunk}.parquet")
```

**Comment by esloch** [comment]
TypeError:
```
  File "/work_lab/fiocruz/repositories/AlertaDengue/AlertaDengue/dbf/sinan.py", line 192, in save_to_pgsql
    row[13] = None if not row[13] else int(row[13])  # nu_idade_n
  File "pandas/_libs/missing.pyx", line 360, in pandas._libs.missing.NAType.__bool__
TypeError: boolean value of NA is ambiguous
``` 
##### Solved with:

>row[13] = (
  >                  None if pd.isnull(row[13]) else int(row[13])
    >            )
```  File "/work_lab/fiocruz/repositories/AlertaDengue/AlertaDengue/dbf/sinan.py", line 198, in save_to_pgsql
    cursor.execute(insert_sql, row)
  File "/docker-env/miniconda/envs/alertadengue/lib/python3.8/site-packages/psycopg2/extras.py", line 146, in execute
    return super().execute(query, vars)
psycopg2.ProgrammingError: can't adapt type 'NAType'
``` 
##### Solved with add new rule:
>row[6] = (
>    None if pd.isnull(row[6]) else str(row[6])
>)
##### Some fields have null values:
``` 

(Pdb) p row
DT_NOTIFIC    2021-07-01
SEM_NOT               26
NU_ANO              2021
DT_SIN_PRI    2021-06-29
SEM_PRI               26
DT_DIGITA     2021-07-05
NM_BAIRRO           <NA>
ID_BAIRRO           None
ID_MUNICIP       2612505
NU_NOTIFIC       3931606
ID_AGRAVO          A92.0
DT_NASC       2021-04-04
CS_SEXO                M
NU_IDADE_N          <NA>

(Pdb) p row.isna()
DT_NOTIFIC    False
SEM_NOT       False
NU_ANO        False
DT_SIN_PRI    False
SEM_PRI       False
DT_DIGITA     False
NM_BAIRRO      True
ID_BAIRRO      True
ID_MUNICIP    False
NU_NOTIFIC    False
ID_AGRAVO     False
DT_NASC       False
CS_SEXO       False
NU_IDADE_N     True

**Review by fccoelho — APPROVED** [review]

**Pull request merged**